decouple openfhe pipeline

AlexanderViand-Intel opened the pull request:
Another openfhe pipeline cleanup motivated by making things nicer for the frontend.

This splits `--mlir-to-openfhe-bgv` into `--mlir-to-bgv --to-openfhe` (and ditto for CKKS).

I'd have liked to do Lattigo, too, but since it doesn't yet support CKKS it'd have been harder to remove the scheme option dependency, and it also does some overriding of the add-client-interface options that the OpenFHE backend doesn't.

On that topic: why *do* we have the option for either per-arg encrypt functions or a single encrypt function? All the OpenFHE E2E tests (and the Frontend) use the former.... If we want to keep the latter, we should (a) make the frontend aware of it so that it generates the right Python helpers and (b) make sure all backends can generate code for either option.

j2kun commented:
> why do we have the option for either per-arg encrypt functions or a single encrypt function

I'd be happy to remove it. I think I added it because, in theory, MLIR supports multiple return values and so you could do it all in one function if you wanted (or in C++ we could return a std::tuple). But I implemented the "do it all in one function" version first and only later decided to split it up for the C++ API.

asraa commented:
Maybe another thing is that the mlir JIT runner that comes with the MLIR tools doesn't support multiple return args, but I suspect when we have a full poly to llvm lowering maybe we just write our own runner.

j2kun commented on tools/heir-opt.cpp:
```diff
@@ -366,18 +366,15 @@ int main(int argc, char **argv) {
-      "to OpenFHE C++ code.",
-      mlirToOpenFheRLWEPipelineBuilder(mlir::heir::RLWEScheme::bgvScheme));
+  PassPipelineRegistration<mlir::heir::EntryFunctionOptions>(
+      "to-openfhe",
```
```suggestion
      "scheme-to-openfhe",
```

I'd prefer a more specific name. At least, using `to-openfhe` makes it harder to find it via grep due to being a subset of other strings (`*-to-openfhe`).

j2kun commented:
> maybe we just write our own runner.

I can't wait for that day because removing the `mlir-cpu-runner` will dramatically speed up build times.

AlexanderViand-Intel commented:
> > maybe we just write our own runner.
> 
> I can't wait for that day because removing the `mlir-cpu-runner` will dramatically speed up build times.

Do we need it to be a *runner*? We're already relying on clang for the E2E OpenFHE stuff, so could we lower to LLVM, emit LLVM IR and then compile instead?

AlexanderViand-Intel commented:
> > why do we have the option for either per-arg encrypt functions or a single encrypt function
> 
> I'd be happy to remove it. I think I added it because, in theory, MLIR supports multiple return values and so you could do it all in one function if you wanted (or in C++ we could return a std::tuple). But I implemented the "do it all in one function" version first and only later decided to split it up for the C++ API.

I think getting rid of the single-function-for-all approach might indeed be best then! 
@ZenithalHourlyRate Is there a specific reason that the Lattigo backend uses the single-function approach or could you change it over to also use per-arg encryption functions?

ZenithalHourlyRate commented:
> I think getting rid of the single-function-for-all approach might indeed be best then!
> @ZenithalHourlyRate Is there a specific reason that the Lattigo backend uses the single-function approach or could you change it over to also use per-arg encryption functions?

I used the single-function approach just because it is handy when writing that pass. Will change it to use the per-arg version.

j2kun commented:
> > > maybe we just write our own runner.
> > 
> > 
> > I can't wait for that day because removing the `mlir-cpu-runner` will dramatically speed up build times.
> 
> Do we need it to be a _runner_? We're already relying on clang for the E2E OpenFHE stuff, so could we lower to LLVM, emit LLVM IR and then compile instead?

Yes, this is what I mean when I say "our own runner": just lower to LLVM and run it for real.

j2kun commented on docs/content/en/docs/Design/simd.md:
```diff
@@ -62,10 +62,10 @@ and index-based accesses into tensors (e.g., `tensor.extract` and
-intended flow: augmenting a high-level program with `secret` annotations, then
-applying the SIMD optimization (and any other high-level optimizations) before
-lowering to BGV operations and then exiting to OpenFHE.
+(e.g., B/FV, BGV, or CKKS). The `--mlir-to-bgv --to-openfhe` pipeline
```
```suggestion
(e.g., B/FV, BGV, or CKKS). The `--mlir-to-bgv --scheme-to-openfhe` pipeline
```

AlexanderViand-Intel replied:
Fixed!

j2kun commented:
@AlexanderViand-Intel can you rebase over main?

AlexanderViand-Intel commented:
> @AlexanderViand-Intel can you rebase over main?

Done! Also renamed the "EntryFunctionOptions" to "OpenfheOptions" as they now include the new "debug" option
